UniStuttgart-VISUS/megamol
# Frontend: Tiling + Explicit View Render Input

Implements **Tiling** for MPI Remote cases (**Powerwall**).
Restructures Image Presentation and View Render Poking to pass a `RenderInput` struct to view event consumption. In the long term we want to pass `RenderInput` to `view.Render()`, but for this we would need to touch all views (again). The curent changes allow to feed arbitrary information to a view. This enables us to tell the view to render in some framebuffer resolution independent of window resolution, as well as pass tiling and camera matrix information to the view.

## Summary of Changes
Restructures `ImagePresentation_Service` and `AbstractView_EventConsumption` to handle and pass around `RenderInput` to views, along with housekeeping of now view-specific render input information.
Introduces `mmSetViewFramebufferSize(string view, int width, int height)` Lua command to set framebuffer resolution for views.
Introduces `--framebuffer` and `--tile` CLI options to set View Framebuffer size and Camera Tile information from CLI, respectively.

## References and Context
Implements Remote Tiling and prepares Camera Matrix injection for VR.

## Bugs, TODO
* **infovis.lua** example not working
    * SplitViewGL not tiling-aware
    * View2DGL Camera State (zoom, pan) not remote synced (via param serialization)
    * Brushing not remote synced
    * Note that remote sync of the graph currently serializes module parameters on the headnode and deserializes them on the rendernode
* **GUI** not tiling-aware?
* **GUI** rendering can not be disabled, leads to bad performance? => restructure for new frontend rendering presentation/blending
* **Screenshots** are still fed from frontbuffer. allow to screenshot ImageWrappers of Entry Point rendering results or some frontend FBO where rendering results got blended into for final presentation.

## Test Instructions
Open your favorite projects and experiment with window-independent view framebuffer resolution via remote console. Note that screenshot are still written from the GL front/backbuffer and not from View Rendering Result Textures yet.

![grafik](https://user-images.githubusercontent.com/44215465/134960383-65196c04-f601-424b-a96f-9af0007f8b9b.png)

Setting up a remote MPI tiling example is more tricky and is provided by the `powerwall_config.lua` configuration file in the `examples/powerwall` directory. Basically you start a bunch of MPI rendernodes with appropriate tiling settings and connect a headnode as remote control. Dont forget to compile your MegaMol with `ENABLE_MPI` activated in CMake and the correct `MPI_GUESS_LIBRARY_NAME` (MSMPI on Windows).

https://user-images.githubusercontent.com/44215465/134998270-71c7a293-b3e2-4dce-84b5-e9fdb6486bdc.mp4

```
Headnode:    ./megamol.exe --headnode -c megamol_config.lua -c ../examples/powerwall/powerwall_config.lua
Rendernodes: mpiexec.exe -n 4 megamol.exe --mpi --rendernode -c megamol_config.lua -c ../examples/powerwall/powerwall_config.lua
```
Beware that when passing config files explicitly to megamol via `-c` you need to pass all config files in the order of evaluation. In this case MegaMol will not implicitly load default config files. The powerwall config file starts a local tiling setup on your machine both for headnode and rendernodes. If the config thinks it runs in the powerwall environment, it will set up a tiling using the powerwall geometry.
